Add chained features API

iliajie opened the pull request:
Hello Jamie!

As discussed [here](https://github.com/virtualmin/virtualmin-gpl/issues/809) but will not only disable dependent features but also re-enable.

jcameron commented on virtual-server-lib-funcs.pl:
```diff
@@ -5836,6 +5836,27 @@ sub quota_parse
+{
+my ($f) = @_;
+my $chained = {
+	mail => ['spam', 'virus'],
```
I wonder if there's other dependencies we could add while we're in there, like webalizer on web ?

iliajie replied:
Sure, then we just add `web => ['webalizer', 'virtualmin-awstats']` ..

jcameron replied:
Although I just realized that they can work with Nginx, so it's not that simple :-(

iliajie replied:
I already fixed that. All good.

jcameron commented on virtual-server-lib-funcs.pl:
```diff
@@ -5836,6 +5836,27 @@ sub quota_parse
+	return "oninput=\"if (!form.$f.checked) { $deps }\"";
+	}
+for my $c (keys %$chained) {
+	return "oninput=\"if (form.$f.checked) ".
```
Is there ever danger that some feature field doesn't exist?

iliajie commented:
I have run more tests and fixed a few bugs. I also added more chained features as you suggested earlier! All feels smooth now. Please have a look https://github.com/virtualmin/virtualmin-gpl/pull/811/commits/2d6ba33d7ec78407a27c9206eab13258b2f887e0.

jcameron commented on virtual-server-lib-funcs.pl:
```diff
@@ -5836,6 +5836,31 @@ sub quota_parse
+	}
+for my $c (keys %$chained) {
+	next if (!$config{$c} && &indexof($c, @plugins) < 0);
+	return "oninput=\"if (form['$f'] && form['$f'].checked) ".
```
Is this changing the checked state of other fields? Shouldn't it be disabling fields that cannot be used?

iliajie replied:
Yes, but it also enables dependent features.

For example, it will disable Spam and Virus if Mail feature is unticked; it will auto-enable Mail if either Spam or Virus is ticked.

jcameron replied:
I think you should just disable dependent features when they cannot be used - going the other way seems quite confusing.

iliajie replied:
I don't think it's confusing at all. On contrary, it doesn't let user choose options that will fail anyway after saving the form.

That said, how it can be confusing selecting AWStats reporting feature and not have Apache/Nginx option enabled?

jcameron replied:
So if the mail feature is disabled, will the checkboxes for spam and virus be disabled? Or just un-checked?

iliajie replied:
> So if the mail feature is disabled, will the checkboxes for spam and virus be disabled? Or just un-checked?

They will be unchecked, i.e. disabled! 😅

jcameron replied:
"disabled" to me means greyed out, not just un-checked

iliajie replied:
Ah, I see. But do we ever gray-out features on **Edit Virtual Server** page? I don't think we do.

This API is made to work only with enabled (not disables) checkboxes. Disabled state is controlled by the server.

jcameron replied:
Ok I see now .. I was confused about what disabling the field meant.

iliajie commented:
Jamie, nearly forgot to add this when creating a new domain! Check this patch please:

https://github.com/virtualmin/virtualmin-gpl/commit/f4433153dbbb017932b9a65f62331575c7b073c4

jcameron commented:
Looks good!
